File: src/#1026/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxAncestorDiff(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return helper(root, root.Val, root.Val)
}

func helper(root *TreeNode, min, max int) int {
	if root == nil {
		return max - min
	}
	if root.Val > max {
		max = root.Val
	}
	if root.Val < min {
		min = root.Val
	}
	left := helper(root.Left, min, max)
	right := helper(root.Right, min, max)
	if left > right {
		return left
	} else {
		return right
	}
}

File: src/#1026/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

var maxAncestorDiff = function (root) {
  if (root === null) {
    return 0;
  }

  return helper(root, root.val, root.val);
  function helper(root, min, max) {
    if (root === null) {
      return max - min;
    }
    if (root.val > max) {
      max = root.val;
    }
    if (root.val < min) {
      min = root.val;
    }
    let left = helper(root.left, min, max);
    let right = helper(root.right, min, max);

    if (left > right) {
      return left;
    } else {
      return right;
    }
  }
};
